Optimize and externalise build tooling

rtibbles opened the pull request:
### Summary
* Makes optimizations to the Kolibri frontend build pipeline by putting linting in a separate process
* Consolidates all frontend linting into a single frontend command
* Creates a `kolibri-tools` npm package using yarn workspaces that exposes commands for linting, building, and testing.
* Makes the `kolibri-tools` package exportable for use by external plugins to build their frontend assets standalone, without requiring a copy of the kolibri source repo (although a pip installed version of Kolibri will still be necessary, for the WebpackFrontendAssetHook import).
* Creates a `kolibri` npm package that creates an (somewhat complete, but with extant issues for some use cases) exported copy of the Kolibri Core API currently only intended for use in tests of external plugins (as any imports of the core API will need to be properly resolved for the purposes of executing tests).
* Simplifies the frontend build logic, consolidating all base webpack configuration generation in a function in the `webpack.config.base.js` that takes a single object with data about the frontend plugin as input, and returns the appropriate webpack config as output.
* Adds an experimental `multi-processor` mode for all build commands. This can be flagged with the `--multi` flag, or by setting `KOLIBRI_BUILD_MULTI=true` as an env var. This will aggressively use all resources available on the machine to attempt to complete the build as fast as possible.
* Upgrades the required node version to 10.
* Deletes the capability to do core API injection, as it was not amenable to the way that this PR refactored the build pipeline. 
* Deletes ununsed `events` and `once` hashes defined on frontend plugins. Has not removed all the associated internal code from the mediator yet, but this should probably also be cleaned up.
* Adds the ability to specify build plugins using globs in Python paths.

### Reviewer guidance
Install Node.js 10.
Make sure that all the build and linting related commands in the package.json still work.
Test out the multi processor option.
Check the logging during builds and linting to see if it meets developer needs.

### References
Fixes #950 
Fixes #3653 
Fixes #4377
Fixes #4443 

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

rtibbles committed:
Put all frontend linting inside prettier-frontend.

rtibbles committed:
Add watch mode to frontend linter.
Remove all linting from webpack.

rtibbles committed:
Paralellize webpack builds.

rtibbles committed:
Cleanup tests.

rtibbles committed:
Move frontend build and other frontend tools into packages.
Setup lerna.js

rtibbles committed:
Add cli for linting, building, and testing.

rtibbles committed:
Reference kolibri logging by core api path.

rtibbles committed:
Use a dynamically generated static path in frontend, rather than static.

rtibbles committed:
Fix up tests and imports.

rtibbles committed:
Fix kolibri-tools test behaviour.

rtibbles committed:
Update linting in intl_code_gen.

rtibbles committed:
Make jest configuration modifiable.

rtibbles committed:
Run a linting pass.

rtibbles committed:
Update gitignore to ignore all node log files.

rtibbles committed:
Complete build pipelines for kolibri and kolibri-tools packages.

rtibbles committed:
Finalize npm publishing config.

rtibbles committed:
Fix clean and bundle-stats behaviours.

rtibbles committed:
Fix api spec exporting and api spec building.

rtibbles committed:
Add fixes to allow building of an external plugin that is also distributed with kolibri.

rtibbles committed:
Simplify and flatten webpack json python generation code.

rtibbles committed:
v0.12.0-dev.1

rtibbles committed:
Remove lerna.
Update building and publishing scripts for packages.

rtibbles committed:
Pipe stdio to parent stdio process in publishing script.
Update kolibri-tools dependency in repo root.
Update yarn.lock.

rtibbles committed:
Consolidate and simplify core webpack build processes.

rtibbles committed:
Add a base eslintrc.js for IDE linter compatibility.

rtibbles committed:
Remove references to old events hashes.
Update travis and docker node version to 10.

rtibbles committed:
Remove test of event hashes functionality.

rtibbles committed:
Fix bug for source vs dist checking for api spec export.

rtibbles committed:
Add system path appending to make sure we can import kolibri.
Update tox env name to represent node upgrade.

aronasorman commented:
![image](https://user-images.githubusercontent.com/191955/48224269-18555580-e34e-11e8-978c-27f58ca98afe.png)

ralphiee22 commented:
Leaving this here as a todo.
Change port number between watching plugin frontend assets and kolibri frontend assets.

rtibbles commented:
@ralphiee22 what I'll do is add an option for 'starting port' for the webpack devservers so that plugins can set a different port by default if they want to, seems like the simplest solution. Work for you?

indirectlylit commented on package.json:
```diff
@@ -31,115 +39,18 @@
-    "webpack-hot-middleware": "2.12.1",
-    "webpack-merge": "^4.1.0",
-    "webpack-rtl-plugin": "^1.6.0",
+    "stylelint-scss": "^3.3.2",
```
why is `stylelint-scss` here and not in `packages/kolibri-tools`?

Also I see `responselike` is used in `__mocks__/http.js` - shouldn't that be in `packages/kolibri-tools` or maybe `kolibri/core`?

indirectlylit commented on packages/publish.js:
```diff
@@ -0,0 +1,75 @@
+
+/*
+ * Step 1 - Set the version of the kolibri package by the current Kolibri version
+ */
```
super minor style point, but would prefer these comments be single-line so this file is easier to scan

also, a block comment at the top describing what this file does would be helpful. What are we publishing and to where?

indirectlylit commented on packages/kolibri/package.json:
```diff
@@ -0,0 +1,35 @@
+    "vuex": "^3.0.1"
+  },
+  "devDependencies": {
+    "kolibri-tools": "0.12.0-dev.1"
```
this is fine because it's an internal dependency, but if we ever make it public I'd prefer to call it `kolibri-dev-tools`

indirectlylit commented on packages/eslint-plugin-kolibri/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Do not ignore this lib dir
```
a 'because' clause would be helpful

indirectlylit commented:
Working on review now. One quick question - how (if at all) will this affect the Perseus dev workflow?

indirectlylit commented on kolibri/core/package.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "kolibri-core",
```
In the notes you say

> Creates a `kolibri` npm package that creates an (somewhat complete, but with extant issues for some use cases) exported copy of the Kolibri Core API currently only intended for use in tests of external plugins (as any imports of the core API will need to be properly resolved for the purposes of executing tests).

I'm a little unclear what this is and how it works.

As an aside, wouldn't the package be `kolibri-core` rather than `kolibri`?

rtibbles replied:
No, the `kolibri` package is constructed with reference to this `kolibri-core` package, but is a separate entity. The main purpose of calling the exported package `kolibri` is for consistency of naming between our internal references to the kolibri core API, and people using it externally.

rtibbles commented:
You will now be able to build Perseus without doing a full build of Kolibri as a whole. So you can build perseus standalone.

indirectlylit commented on packages/kolibri-tools/lib/cli.js:
```diff
@@ -0,0 +1,297 @@
+      }
+      const children = [];
+      for (let index = 0; index < numberOfBundles; index++) {
+        if (multi) {
```
This works well on my machine! Building takes noticeably shorter, and makes my computer noticeably hotter :)

If you wanted to prevent all threads from kicking off simultaneously, you could maybe keep a count `numBuildsRunning` and then do something like

```javascript
function build() { /* create child process */ }

function buildEventually() {
  if (numBuildsRunning > numCores - 1) {
    setTimeout(75, buildEventually);
  } else {
    build();
  }
}

buildEventually();
```

indirectlylit replied:
(as we discussed, this wouldn't work: I didn't realize the processes stay running rather than restart on each rebuild)

indirectlylit commented on kolibri/core/assets/src/kolibri_module.js:
```diff
@@ -25,9 +25,6 @@ export default class KolibriModule {
     this.name = __kolibriModuleName;
-    // Insert the event hashes defined in the Python code.
-    this.events = __events;
-    this.once = __once;
```
yay!

indirectlylit commented on kolibri/core/assets/src/views/KGrid/extra-units.css:
```diff
@@ -78,7 +77,7 @@
 
 .pure-u-7-7 {
-  width: 100.0%;
+  width: 100%;
```
I don't know for sure whether the extra precision was originally necessary - it's based on pure.css grids and they do some fairly tricky browser compatibility stuff. I'd prefer to ignore the file from linting and keep the decimal precision as-is to be on the safe side if possible.

rtibbles replied:
Righto, can also add linting exclude comments.

indirectlylit reviewed (approved):
Looks great! I like it a lot.

Minor comments and questions, but overall nothing blocking

indirectlylit commented:
note on logging: when the threads are parallelized, some of the messages make less sense because they are interleaved:

```
Kolibri: Completed device_management_side_nav bundle in 4.138s!
Kolibri: Completed user_module_user_profile_nav_side_nav bundle in 4.209s!
ℹ ｢wdm｣:    12 modules
ℹ ｢wdm｣: Compiled successfully.
ℹ ｢wdm｣:    12 modules
ℹ ｢wdm｣: Compiled successfully.
Kolibri: Completed coach_side_nav bundle in 4.108s!
Kolibri: Completed user_module_login_nav_side_nav bundle in 4.084s!
ℹ ｢wdm｣:    12 modules
ℹ ｢wdm｣: Compiled successfully.
Kolibri: Completed learn_module_side_nav bundle in 3.961s!
ℹ ｢wdm｣:    12 modules
ℹ ｢wdm｣: Compiled successfully.
ℹ ｢wdm｣:    11 modules
ℹ ｢wdm｣: Compiled successfully.
Kolibri: Completed facility_management_side_nav bundle in 4.227s!
ℹ ｢wdm｣:    12 modules
ℹ ｢wdm｣: Compiled successfully.
```

Might as well just hide the ones with no context completely

indirectlylit committed:
first pass of HMR support

indirectlylit committed:
add parameter for making it hot

indirectlylit committed:
skip plugins when it's not hot

indirectlylit commented on packages/kolibri-tools/yarn-error.log:
```diff
@@ -0,0 +1,14515 @@
+  /home/richard/.nvm/versions/node/v10.12.0/bin/node /home/richard/.yarn/bin/yarn.js add webpack-file-loader
+
+PATH: 
+  /home/richard/.nvm/versions/node/v10.12.0/bin:/home/richard/.virtualenvs/kolibripy2/bin:/home/richard/.yarn/bin:/home/richard/.config/yarn/global/node_modules/.bin:/usr/local/bin:/home/richard/.cargo/bin:/home/richard/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/lib/jvm/java-8-oracle/bin:/usr/lib/jvm/java-8-oracle/db/bin:/usr/lib/jv ...
```
whoops - looks like this error log was committed by accident. My current PR has some yarn issues, too. I'll fix those those up once this is updated

indirectlylit replied:
nevermind - addressed it here: https://github.com/rtibbles/kolibri/pull/16/commits/62932beb37bc2cf89ccfc8dd228817bc585d2307

indirectlylit committed:
update dependencies

indirectlylit committed:
add missing version number in demo_server/package.json

indirectlylit committed:
bump vue-intl version and switches build we use - prod vs dev

Previously we were always using the dev version. This change relies on https://github.com/learningequality/vue-intl/pull/22

indirectlylit commented on package.json:
```diff
@@ -1,25 +1,33 @@
+    "lint:watch": "yarn lint:fix -m",
+    "generate-locale-data": "node ./build_tools/i18n/intl_code_gen.js",
+    "build-kolibri": "yarn workspace kolibri run build",
+    "build-kolibri-tools": "yarn workspace kolibri-tools run build",
```
I'm unclear on what `build-kolibri` and `build-kolibri-tools` are for, and how they differ from `build` above.

`build-kolibri` fails with 

```
(k012x)d@mac ~/P/l/r/k012x (external_build)> yarn run build-kolibri
yarn run v1.12.3
$ yarn workspace kolibri run build
warning Missing version in workspace at "/Users/d/Projects/learning_equality/repos/k012x/kolibri/plugins/demo_server", ignoring.
$ node ./build/index.js
internal/modules/cjs/loader.js:582
    throw err;
    ^

Error: Cannot find module '/Users/d/Projects/learning_equality/repos/k012x/packages/kolibri/build/index.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:580:15)
    at Function.Module._load (internal/modules/cjs/loader.js:506:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:741:12)
    at startup (internal/bootstrap/node.js:285:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:739:3)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1
Command: /Users/d/.nvm/versions/node/v10.14.1/bin/node
Arguments: /usr/local/Cellar/yarn/1.12.3/libexec/lib/cli.js run build
Directory: /Users/d/Projects/learning_equality/repos/k012x/packages/kolibri
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

rtibbles replied:
These are the build commands for the two exportable/publishable npm modules.

rtibbles replied:
I think I must have committed this code part way through a refactor of how I was doing the building...

rtibbles replied:
I have building working again, but the `kolibri` package is still very much WIP.

indirectlylit commented on package.json:
```diff
@@ -1,25 +1,33 @@
-    "test-build": "yarn run lint-js && webpack --config ./frontend_build/src/webpack.config.prod.js --bail",
-    "makemessages": "webpack --config ./frontend_build/src/webpack.config.trs.js",
-    "watch": "node ./frontend_build/src/webpackdevserver.js",
+    "build": "kolibri-tools build prod --file ./build_tools/build_plugins.txt",
```
The build command currently clears the console after each module is built. It would be better if it didn't, because helpful information such as bundle sizes gets hidden.

(Note - I'm using `multi` mode)

rtibbles replied:
This happens without multi also. Will have to look and see how to stop webpack from doing this.

indirectlylit commented on packages/kolibri-tools/lib/cli.js:
```diff
@@ -0,0 +1,297 @@
+    list,
+    []
+  )
+  .option('-m, --multi', 'Run using multiple cores to improve build speed', false)
```
Might want to make the default `true`?

indirectlylit reviewed (commented):
a few more comments based on recent HMR work

indirectlylit committed:
update devserver getting-started docs

indirectlylit commented on .eslintrc.js:
```diff
@@ -1,124 +1 @@
-    'kolibri/vue-component-registration-casing': ERROR,
-  },
-};
+module.exports = require('kolibri-tools/.eslintrc');
```
As far as I can tell, vue files are no longer being linted after this change. This is true both of `<script>` and `<template>` sections.

To test this, try adding `props: ['simpleprop']` to a component before and after this PR. Before, it will complain

```
eslint:vue/require-prop-types Props should at least define their types.
```

but after it will not.

Currently I believe this is a config issue and not related to `lint.js` because the behavior happens whether I run the `devserver` or try to set up linting in sublimetext.

Investigating further...

indirectlylit replied:
Confirmed that moving `.eslintrc.js` back to the root directory and updating the `import/resolver` allows `eslint` to work correctly in Sublime.

However, even with this change, `yarn run lint` behaves strangely:

* `<template>` errors are thrown without context, e.g. `Kolibri linter: Invalid array length` (which I believe comes from `HTMLHint`, not `eslint`)
* Vue style rules like `require-prop-types` are not reported
* Syntax errors in the `<script>` section are reported, but again lack context (filename)

rtibbles replied:
I hate magic import resolvers. Will play around with this and try to add some testing.

Will look at the context issues again, I was trying to use the built in reporters for each linter to maintain context as much as possible, but it's possible that invoking them with the file contents and not the file itself may be causing issues. Thanks for checking this out!

rtibbles replied:
I have tested this out, and this all seems to be working properly now, and eslint is working correctly in Sublime for me also. Let me know if this is still causing issues as the PR is now.

indirectlylit replied:
working!

indirectlylit reviewed (dismissed):
noticed a regression

indirectlylit committed:
clean up and prettify

indirectlylit commented on packages/kolibri-tools/lib/cli.js:
```diff
@@ -0,0 +1,297 @@
+      [modes.DEV]: 'webpackdevserver.js',
+      [modes.I18N]: 'i18n.js',
+      [modes.STATS]: 'bundleStats.js',
+    }[mode];
```
This currently returns `undefined` when running `make clean`, which causes the line below to error out

rtibbles replied:
Have brought clean in line with the other modes.

rtibbles committed:
Merge pull request #16 from indirectlylit/external_build

make it hot

rtibbles committed:
Fix up simple build related tweaks.

rtibbles committed:
Fix eslint and prettier linting and error reporting.
Lint files.

rtibbles committed:
Update linting for styling, including autofixing. Lint files. Update watcher command.

rtibbles committed:
Add custom port option for devserver and bundlestats.

rtibbles committed:
Merge branch 'develop' into external_build

rtibbles committed:
Update node 10 version.

rtibbles committed:
Fix up stylint lint handling to account for differential behaviour when
style code is parseable or not.

rtibbles committed:
Add extra check and code comment.

indirectlylit reviewed (approved):
great job
